include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -0,0 +1,1386 @@
+	} CASE_END
+#endif
+#ifdef OP_BLEND
+	CASE_BEGIN(op, OP_BLEND) {
+		result = instr_execute_cos_eval_pdf_BLEND(instr, srcs, params, bsdf_data);
+	} CASE_END
+#endif
+#ifdef OP_BUMPMAP
+	CASE_BEGIN(op, OP_BUMPMAP) {
+		instr_execute_BUMPMAP(instr, srcs);
+	} CASE_END
+#endif
+#ifdef OP_SET_GEOM_NORMAL
+	CASE_BEGIN(op, OP_SET_GEOM_NORMAL) {
+		instr_execute_SET_GEOM_NORMAL();
+	} CASE_END
+#endif
+	CASE_OTHERWISE
+	{} //else "empty braces"
+	END_CASES
+
+	if (op_isBXDForBlend(op))
+		writeReg(REG_DST(regs), result);
+}
+
+eval_and_pdf_t irr_bsdf_eval_and_pdf(in instr_stream_t stream, inout irr_glsl_LightSample s, in instr_t generator, inout irr_glsl_AnisotropicMicrofacetCache uf)
+{
+	bool ts = false;
+	const bool is_generator_bsdf = op_isBRDF(instr_getOpcode(generator));
+	instr_execute_SET_GEOM_NORMAL();
+	for (uint i = 0u; i < stream.count; ++i)
+	{
+		instr_t instr = irr_glsl_MC_fetchInstr(stream.offset+i);
+		uint op = instr_getOpcode(instr);
+
+		if (INSTR_1ST_DWORD(instr) != INSTR_1ST_DWORD(generator))
+		{
+			instr_eval_and_pdf_execute(instr, s, uf, ts);
+		}
+		else
+		{
+			//assert(isBXDF(instr));
+			uvec3 regs = instr_decodeRegisters(instr);
+			writeReg(REG_DST(regs), eval_and_pdf_t(0.0));
+		}
+
+#if defined(OP_SET_GEOM_NORMAL)||defined(OP_BUMPMAP)
+		if (
+#ifdef OP_SET_GEOM_NORMAL
+			op == OP_SET_GEOM_NORMAL
+#ifdef OP_BUMPMAP
+			||
+#endif
+#endif
+#ifdef OP_BUMPMAP
+			op == OP_BUMPMAP
+#endif
+			) {
+			s = irr_glsl_createLightSample(s.L, currInteraction);
+			uf = irr_glsl_calcAnisotropicMicrofacetCache(currInteraction, s);
+		}
+#endif
+	}
+
+	eval_and_pdf_t eval_and_pdf = readReg4(0u);
+	return eval_and_pdf;
+}
+
+irr_glsl_AnisotropicMicrofacetCache getSmoothMicrofacetCache(in float NdotV, in float NdotL)
+{
+	irr_glsl_AnisotropicMicrofacetCache uf;
+	const float d = 1e-4;//some delta to avoid zeroes... not a perfect solution but for now..
+	uf.isotropic.VdotH = NdotV-d;
+	uf.isotropic.LdotH = NdotL-d;
+	uf.isotropic.NdotH = d;
+	uf.isotropic.NdotH2 = d*d;
+	uf.TdotH = 1.0-d;
+	uf.BdotH = 1.0-d;
+
+	return uf;
+}
+
+irr_glsl_LightSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand, out vec3 out_remainder, out float out_pdf, out instr_t out_generatorInstr, out irr_glsl_AnisotropicMicrofacetCache out_uf)
+{
+	uint ix = 0u;
+	instr_t instr = irr_glsl_MC_fetchInstr(stream.offset);
+	uint op = instr_getOpcode(instr);
+	float weight = 1.0;
+	vec3 u = rand;
+
+	instr_execute_SET_GEOM_NORMAL();
+	while (!op_isBXDF(op))
+	{
+#ifdef OP_BLEND
+		if (op==OP_BLEND) {
+			bsdf_data_t bsdf_data = fetchBSDFDataForInstr(instr);
+			params_t params = instr_getParameters(instr, bsdf_data);
+			float w = params_getBlendWeight(params);
+			float rcpChoiceProb;
+			bool choseRight = irr_glsl_partitionRandVariable(w, u.z, rcpChoiceProb);
+
+			uint right = instr_getRightJump(instr);
+			ix = choseRight ? right:(ix+1u);
+			weight *= choseRight ? (1.0-w):w;
```

Comment:
you could just divide `out_pdf` by `rcpChoiceProb` or multiply it here instead of `weight`, as long as `out_pdf` was initialized to 1.0